### DISCUSSION TITLE
Next v13.4 / App Router / how to make server component list blogs on server side and then user can select filter to filter blogs on client side

### DISCUSSION THREAD

**Discussion by shirishmakwana** [vercel/next.js#53834] [Help]
### Summary

I am new to next js.

Let me explain with example I am trying.
There is blogs page. I want to list blogs in server side component. Then there are some filter option available on the page above the blog listing. Suppose these are past and upcoming. So wanted to filter blogs client side with selected filter. How to write this serve component and filter component
Thank you.

### Additional information

```js
Next V13.4
App Router
Using src folder
```

**Comment by masterkain** [comment]
I don't suggest doing this client side, the idea is to create links that aggregate query params that when clicked they will recall the blogs url with updated params, and then in your server side page component consult those params and filter them before display.

**Reply by shirishmakwana** [reply]
Thank you for your prompt response @masterkain 

Indeed, you're correct. While considering the scenario of having blogs, let's say I intend to display a list of events for the entire year (let's assume a maximum of 20 events for the year). My objective is to efficiently filter these events, distinguishing between upcoming and past events. I'm interested in understanding the most effective approach to achieve this, both from the perspective of client-side and server-side implementation. Your insights would be greatly appreciated.

**Reply by masterkain** [reply]
https://github.com/icoretech/airbroke/blob/d9df921a39b7d85c2b0841eb8c57a55b927b07ee/app/projects/page.tsx#L18
here in our page we fetch the current parameters, and we send them down to the ProjectsTable further down the file.

then in the actual table you'll find a utility function that actually takes what it needs from the params and returns the filtered records:
https://github.com/icoretech/airbroke/blob/main/components/ProjectsTable.tsx#L16

in this case mine is `getProjects` that will query a database via prisma for a search term.

---

when you need to concatenate multiple params for further filtering you can use some utils such as this one generateUpdatedURL
https://github.com/icoretech/airbroke/blob/main/lib/generateUpdatedUrl.ts
you can check it in action in the Filter component, which should be near what you're trying to achieve: https://github.com/icoretech/airbroke/blob/d9df921a39b7d85c2b0841eb8c57a55b927b07ee/app/projects/%5Bproject_id%5D/Filter.tsx#L4

**Reply by shirishmakwana** [reply]
Thank you. Checked above suggestions.
Currently, with each alteration made to the filter, the page is fetched anew from the server. This activity is visible within the network tab. Is there a method to address and manage this behavior as well?

**Reply by masterkain** [reply]
that's correct up until a certain point, you see -- next.js has many caching layers, some can be controlled, some do not, some are `fetch` dependent, etc. so it's better the check the updated caching guide https://nextjs.org/docs/app/building-your-application/caching

then if you fear that requesting the page anew from the server will cause flickering I suggest making sure you wire up things correctly first because there should be no flickering and it should be almost as if the component rerenders in browser, if you know what I mean

**Reply by Xanacas** [reply]
If I understand you correct, you want the server component to provide all the events/entries to a list and then you would like to give the user the ability to filter through them while doing the filter only client side?
If so, I'd query the entries from the database in a server component and just pass them down into a client-component as prop. your clientcomponent handles then the rendering and filter client side.

**Reply by shirishmakwana** [reply]
@masterkain - That's precisely what I'm investigating: the possibility of achieving the filtering without involving the server.

@Xanacas - You've captured the essence of the question. While we can render the list through props on the client component, I'm curious about the initial rendering of the list on the server side. How can we approach rendering it on the server side first and then proceed to implement the filtering on the client side?

**Reply by Xanacas** [reply]
@shirishmakwana as far as I know, also Clientcomponents are rendered serverside, if you do not navigate clientside to the page. But I’m not sure about that. But as I’ve learned also client components are indexable by search engines, they must be rendered serverside at least in certain conditions.

If for some reasons you want to make sure, it’s a true server component on initial render. 
you could fetch data in parent and the forward via props to either a server component (if no filter is applied ) or conditionally to a client component. However in that case you would need to put the filter as searchparams to the url.
